src/cleanfigure.m

```diff
@@ -114,16 +114,32 @@ function cleanfigure(varargin)
           movePointsCloser(meta, h);
           % Don't be too precise.
           coarsenLine(meta, h, minimumPointsDistance);
+
       elseif strcmp(type, 'text')
-          % Check if text is inside bounds by checking if the Extent rectangle
-          % and the axes box overlap.
-          xlim = get(meta.gca, 'XLim');
-          ylim = get(meta.gca, 'YLim');
+          % Check if text is inside bounds by checking if the position is inside
+          % the x, y and z limits. This works for both 2D and 3D plots.
+          x_lim = get(meta.gca, 'XLim');
+          y_lim = get(meta.gca, 'YLim');
+          z_lim = get(meta.gca, 'ZLim');
+          axLim = [x_lim; y_lim; z_lim];
+
+          pos = get(h, 'Position');
+          bPosInsideLim = ( pos' >= axLim(:,1) ) & ( pos' <= axLim(:,2) );
+
+          % In 2D plots the extent of the textbox is available and also
+          % considered to keep the textbox, if it is partially inside the axis
+          % limits.
           extent = get(h, 'Extent');
-          extent(3:4) = extent(1:2) + extent(3:4);
-          overlap = xlim(1) < extent(3) && xlim(2) > extent(1) ...
-                 && ylim(1) < extent(4) && ylim(2) > extent(2);
-          if ~overlap
+          if all(~isnan(extent))
```

Comment:
Please open a disussion issue for how to proceed with textboxes outside a single axis. I also wondered why those should be removed, but this is the way the code handled the situation before I laid my hands on it. Shortly I will add another commit that adds a comment.

The [help](http://de.mathworks.com/help/matlab/ref/text-properties.html?searchHighlight=extent) states that by default `extent` is defined in `data units` but it can be overwritten with the property `Units`.